destination/writer/writer.go

```diff
@@ -0,0 +1,317 @@
+	}
+
+	return nil
+}
+
+// generates an SQL INSERT statement query.
+func (w *Writer) buildInsertQuery(table string, columns []string, values []any) (string, error) {
+	if len(columns) != len(values) {
+		return "", errColumnsValuesLenMismatch
+	}
+
+	ib := sqlbuilder.NewInsertBuilder()
+
+	ib.InsertInto(table)
+	ib.Cols(columns...)
+	ib.Values(values...)
+
+	sql, args := ib.Build()
+
+	query, err := sqlbuilder.DefaultFlavor.Interpolate(sql, args)
+	if err != nil {
+		return "", fmt.Errorf("interpolate arguments to SQL: %w", err)
+	}
+
+	return query, nil
+}
+
+// generates an SQL UPDATE statement query.
+func (w *Writer) buildUpdateQuery(
+	table string, keyColumn string, keyValue any, columns []string, values []any,
+) (string, error) {
+	if len(columns) != len(values) {
+		return "", errColumnsValuesLenMismatch
+	}
+
+	ub := sqlbuilder.NewUpdateBuilder()
+
+	ub.Update(table)
+	ub.Where(
+		ub.Equal(keyColumn, keyValue),
+	)
+
+	assignments := make([]string, len(columns))
+	for i := 0; i < len(columns); i++ {
+		assignments[i] = ub.Assign(columns[i], values[i])
+	}
+	ub.Set(assignments...)
+
+	query, err := sqlbuilder.DefaultFlavor.Interpolate(ub.Build())
+	if err != nil {
+		return "", fmt.Errorf("interpolate arguments to SQL: %w", err)
+	}
+
+	return query, nil
+}
+
+// generates an SQL DELETE statement query.
+func (w *Writer) buildDeleteQuery(table string, keyColumn string, keyValue any) (string, error) {
+	db := sqlbuilder.NewDeleteBuilder()
+
+	db.DeleteFrom(table)
+	db.Where(
+		db.Equal(keyColumn, keyValue),
+	)
+
+	query, err := sqlbuilder.DefaultFlavor.Interpolate(db.Build())
+	if err != nil {
+		return "", fmt.Errorf("interpolate arguments to SQL: %w", err)
+	}
+
+	return query, nil
+}
+
+// returns either the record metadata value for the table
+// or the default configured value for the table.
+func (w *Writer) getTableName(metadata map[string]string) string {
+	tableName, ok := metadata[metadataTable]
+	if !ok {
+		return w.cfg.Table
+	}
+
+	return strings.ToLower(tableName)
+}
+
+// returns either the first key within the Key structured data
+// or the default key configured value for key.
+func (w *Writer) getKeyColumn(key sdk.StructuredData) (string, error) {
+	if len(key) > 1 {
+		return "", errCompositeKeysNotSupported
+	}
+
+	for k := range key {
+		return k, nil
+	}
+
+	return w.cfg.KeyColumn, nil
+}
+
+// converts sdk.Data to sdk.StructuredData.
+func (w *Writer) structurizeData(data sdk.Data) (sdk.StructuredData, error) {
+	if data == nil || len(data.Bytes()) == 0 {
+		return nil, nil
+	}
+
+	structuredData := make(sdk.StructuredData)
+	if err := json.Unmarshal(data.Bytes(), &structuredData); err != nil {
+		return nil, fmt.Errorf("unmarshal data into structured data: %w", err)
+	}
+
+	return structuredData, nil
+}
+
+// turns the payload into slices of columns and values for use in queries to Oracle.
+func (w *Writer) extractColumnsAndValues(payload sdk.StructuredData) ([]string, []any) {
+	var (
```

Comment:
maybe change it to
```
	var (
		columns = make([]string,  len(payload))
		values  = make([]any,  len(payload))
		i  :=  0
	)
        
	for key, value := range payload {
		columns[i] = key
		values[i] = value
		i = i +1
	}
```

Reply:
I thought about it, but the benchmark said that it is the same time for the one operation.

Reply:
```
var (
		columns []string
		values  []any
	)

	for key, value := range payload {
		columns = append(columns, key)
		values = append(values, value)
	}
```

`BenchmarkWriter_extractColumnsAndValues-8   	 2300732	       495.2 ns/op`

===

```
var (
		columns = make([]string, 0, len(payload))
		values  = make([]any, 0, len(payload))
	)

	for key, value := range payload {
		columns = append(columns, key)
		values = append(values, value)
	}
```

`4749811	       232.5 ns/op`

===

```
var (
		i = 0

		columns = make([]string, len(payload), len(payload))
		values  = make([]any, len(payload), len(payload))
	)

	for key, value := range payload {
		columns[i] = key
		values[i] = value

		i++
	}
```

`4355790	       232.7 ns/op`

Reply:
The longer map or slice the more difference between options

```
func BenchmarkTestOption3(b *testing.B) {
	b.StopTimer()
	m := generateMap()
	b.StartTimer()
	for i := 0; i < b.N; i++ {
		Option3(m)
	}
}

func BenchmarkTestOption2(b *testing.B) {
	b.StopTimer()
	m := generateMap()
	b.StartTimer()
	for i := 0; i < b.N; i++ {
		Option2(m)
	}
}

func generateMap() map[int]any {
	m := make(map[int]any)
	for i := 0; i < 100000; i++ {
		m[i] = i
	}

	return m
}

///////////////////////
func Option3(payload map[int]any) {
	var (
		i = 0

		columns = make([]int, len(payload), len(payload))
		values  = make([]any, len(payload), len(payload))
	)

	for key, value := range payload {
		columns[i] = key
		values[i] = value

		i++
	}
}

func Option2(payload map[int]any) {
	var (
		columns = make([]int, 0, len(payload))
		values  = make([]any, 0, len(payload))
	)

	for key, value := range payload {
		columns = append(columns, key)
		values = append(values, value)
	}
}

```

Reply:
The longer map or slice the bigger difference
```
func Option3(payload map[int]any) {
	var (
		i = 0

		columns = make([]int, len(payload), len(payload))
		values  = make([]any, len(payload), len(payload))
	)

	for key, value := range payload {
		columns[i] = key
		values[i] = value

		i++
	}
}

func Option2(payload map[int]any) {
	var (
		columns = make([]int, 0, len(payload))
		values  = make([]any, 0, len(payload))
	)

	for key, value := range payload {
		columns = append(columns, key)
		values = append(values, value)
	}
}


////////


import (
	"testing"
)

func BenchmarkTestOption3(b *testing.B) {
	b.StopTimer()
	m := generateMap()
	b.StartTimer()
	for i := 0; i < b.N; i++ {
		Option3(m)
	}
}

func BenchmarkTestOption2(b *testing.B) {
	b.StopTimer()
	m := generateMap()
	b.StartTimer()
	for i := 0; i < b.N; i++ {
		Option2(m)
	}
}

func generateMap() map[int]any {
	m := make(map[int]any)
	for i := 0; i < 100000; i++ {
		m[i] = i
	}

	return m
}

```

Reply:
10000 columns? :)
Ok, I'll double-check.

Reply:
it is not about  columns, it is only like good practice .

Reply:
Ok, I'll update it.

Reply:
Done.